### PR TITLE
File locking

### PR DESCRIPTION
**Pull request by vpodzime**
Merge together:
https://github.com/cfengine/core/pull/3880
https://github.com/cfengine/enterprise/pull/556
https://github.com/cfengine/libntech/pull/27

### CODE REVIEW

File: libutils/file_lib.c [hunk 1]
```diff
@@ -1563,6 +1563,28 @@ int ExclusiveFileLock(FileLock *lock, bool wait)
     return LockFD(lock->fd, F_WRLCK, wait);
 }
 
+int ExclusiveFileLockPath(FileLock *lock, const char *fpath, bool wait)
+{
+    assert(lock != NULL);
+    assert(lock->fd < 0);
```
Comment (olehermanse) [inline libutils/file_lib.c]: There's quite a few places where you assign `lock->fd` to -1 or compare it to 0. I would add a comment or use the macro (`EMPTY_FILE_LOCK`) to make it more obvious what this means.

File: libutils/logging.c [hunk 2]
```diff
@@ -273,7 +275,34 @@ const char *GetErrorStr(void)
...
+#else
+
+const char *GetErrorStrFromCode(int error_code)
+{
+    static char errbuf[CF_BUFSIZE];
+    int len;
+
+    memset(errbuf, 0, sizeof(errbuf));
+
+    if (FormatMessage(FORMAT_MESSAGE_FROM_SYSTEM, NULL, error_code,
```
Comment (olehermanse) [inline libutils/logging.c]: ❤️ 💟 ❤️ 💟 ❤️ 💟

File: libutils/logging.c [hunk 3]
```diff
@@ -273,7 +275,34 @@ const char *GetErrorStr(void)
...
+                  MAKELANGID(LANG_ENGLISH, SUBLANG_ENGLISH_US), errbuf, CF_BUFSIZE, NULL))
+    {
+        // remove CRLF from end
+        len = strlen(errbuf);
+        errbuf[len - 2] = errbuf[len - 1] = '\0';
```
Comment (olehermanse) [inline libutils/logging.c]: What if `len` is `0`?

File: libutils/file_lib.c (hunk 1 above)
Comment (vpodzime) [reply libutils/file_lib.c]: It's an invalid file descriptor, what's pretty clear :wink: Yeah, I'll add a note that it's used as an indicator if the lock is initialized or not.

File: libutils/logging.c (hunk 3 above)
Comment (vpodzime) [reply libutils/logging.c]: No idea, I'm not touching this code :smile:

Comment (vpodzime) [reply libutils/logging.c]: (just moved it around)

**Pull request merged**